File: sdks/nodejs-agent/src/clients/BookingClient.ts
import { Config } from '../Config';
import { TransportInterface } from '../transport/TransportInterface';
import { BookingCreate } from '../DTO/BookingCreate';

export class BookingClient {
  constructor(
    private readonly transport: TransportInterface,
    private readonly config: Config
  ) {}

  public async create(dto: BookingCreate, idempotencyKey?: string): Promise<Record<string, unknown>> {
    const payload = dto.toArray();
    if (!payload.agreement_ref) {
      throw new Error('agreement_ref required');
    }
    if (!payload.supplier_id) {
      throw new Error('supplier_id required');
    }
    return this.transport.bookingCreate(payload, idempotencyKey);
  }

  public async modify(
    supplierBookingRef: string,
    fields: Record<string, unknown>,
    agreementRef: string,
    sourceId?: string
  ): Promise<Record<string, unknown>> {
    return this.transport.bookingModify({
      supplier_booking_ref: supplierBookingRef,
      agreement_ref: agreementRef,
      fields,
    });
  }

  public async cancel(
    supplierBookingRef: string,
    agreementRef: string,
    sourceId?: string
  ): Promise<Record<string, unknown>> {
    return this.transport.bookingCancel({
      supplier_booking_ref: supplierBookingRef,
      agreement_ref: agreementRef,
    });
  }

  public async check(
    supplierBookingRef: string,
    agreementRef: string,
    sourceId?: string
  ): Promise<Record<string, unknown>> {
    return this.transport.bookingCheck(supplierBookingRef, agreementRef, sourceId);
  }
}

File: sdks/go-agent/booking_client.go
package sdk

import (
	"context"
	"fmt"
)

// BookingClient provides booking management functionality
type BookingClient struct {
	transport Transport
	config    *Config
}

// NewBookingClient creates a new BookingClient
func NewBookingClient(transport Transport, config *Config) *BookingClient {
	return &BookingClient{
		transport: transport,
		config:    config,
	}
}

// Create creates a new booking
func (bc *BookingClient) Create(ctx context.Context, booking *BookingCreate, idempotencyKey string) (*BookingResult, error) {
	if booking == nil {
		return nil, fmt.Errorf("booking required")
	}

	if booking.AgreementRef == "" {
		return nil, fmt.Errorf("agreement_ref required")
	}

	if booking.SupplierID == "" {
		return nil, fmt.Errorf("supplier_id required")
	}

	payload := booking.ToMap()
	result, err := bc.transport.BookingCreate(ctx, payload, idempotencyKey)
	if err != nil {
		return nil, err
	}

	return BookingResultFromMap(result), nil
}

// Modify modifies an existing booking
func (bc *BookingClient) Modify(ctx context.Context, supplierBookingRef string, fields map[string]interface{}, agreementRef string, sourceID string) (*BookingResult, error) {
	if supplierBookingRef == "" {
		return nil, fmt.Errorf("supplier_booking_ref required")
	}

	if agreementRef == "" {
		return nil, fmt.Errorf("agreement_ref required")
	}

	if fields == nil {
		fields = make(map[string]interface{})
	}

	payload := map[string]interface{}{
		"supplier_booking_ref": supplierBookingRef,
		"agreement_ref":        agreementRef,
		"fields":               fields,
	}

	result, err := bc.transport.BookingModify(ctx, payload)
	if err != nil {
		return nil, err
	}

	return BookingResultFromMap(result), nil
}

// Cancel cancels a booking
func (bc *BookingClient) Cancel(ctx context.Context, supplierBookingRef, agreementRef, sourceID string) (*BookingResult, error) {
	if supplierBookingRef == "" {
		return nil, fmt.Errorf("supplier_booking_ref required")
	}

	if agreementRef == "" {
		return nil, fmt.Errorf("agreement_ref required")
	}

	payload := map[string]interface{}{
		"supplier_booking_ref": supplierBookingRef,
		"agreement_ref":        agreementRef,
	}

	result, err := bc.transport.BookingCancel(ctx, payload)
	if err != nil {
		return nil, err
	}

	return BookingResultFromMap(result), nil
}

// Check checks the status of a booking
func (bc *BookingClient) Check(ctx context.Context, supplierBookingRef, agreementRef, sourceID string) (*BookingResult, error) {
	if supplierBookingRef == "" {
		return nil, fmt.Errorf("supplier_booking_ref required")
	}

	if agreementRef == "" {
		return nil, fmt.Errorf("agreement_ref required")
	}

	result, err := bc.transport.BookingCheck(ctx, supplierBookingRef, agreementRef, sourceID)
	if err != nil {
		return nil, err
	}

	return BookingResultFromMap(result), nil
}

